README.md
wagesOfSin
==========


changeLog
3/26/2013 Jet Goodson <jetgoodson CHEZ gmail.com>
	* moved prameters to wages of sin for randomOrchard.R
	

3/20/2013 Jet Goodson <jetgoodson CHEZ gmail.com>
	A cogInTheMachine.R
	- playing around with SVM - commented out

3/16/2013 Jet Goodson <jetgoodson CHEZ gmail.com>
	M changeLog
	M wagesOfSin.R
	M toolBox.R
	A randomOrchard.R
	* filled in rest of framework for predicting and testing
	* data is reduced now to just categories and contract time/type for prelim run
	* added some plots; 'cuz we're physicists

3/14/2013 Jet Goodson <jetgoodson CHEZ gmail.com>
	M changeLog
	- got rid of  @s since the git repo is public to spam spiders
        M toolBox.R
	* changed, updated, and expanded text cleaning
	
3/8/2013 Katie Malone <cmmalone.158 CHEZ gmail.com>
    - cloned directory and tried test push
3/7/2013 Jet Goodson <jetgoodson CHEZ gmail.com>
	A wagesOfSin.R
	A lojack.R
	A wordBag.R
	A STRATEGI_2010_GAZETTEER.txt 
	A toolBox.R
	A Location_Tree.csv
	A changeLog
	* lojack.R is for handling location data in separate sources
	* wordBag.R is for the word bag and wordbag type analysis
	* moved data files to data folder


lojack.R
#################################
# lojack.R
# written by Jet Goodson 
# started on 7 March 2013
# contact at jetgoodson@gmail.com
#
# toolset for building a place name
# to lat/long coords look up table
# and other geospatial tools
##################################

##Contains Ordnance Survey data © Crown copyright and database right 2010
#Contains Royal Mail data © Royal Mail copyright and database right 2010
#Contains National Statistics data © Crown copyright and database right 2010


#native test function for testing other functions
lojack <- function() {

  coordinates <- createCoordinateTable()
  print(head(coordinates))
  locations <- createLocationTable()
  print(head(locations))
 
  for(i in 1:nrow(locations)){
     
    result <- coordinateCheck(locations[i,])
    cat("got results \n")
    if(!(result == "okay")){
     cat(c(locations[i,2], "~", locations[i,3], "~", locations[i,4], "~", locations[i,5], "~", locations[i,6], "~", locations[i,7], "~", locations[i,8], " ~~~ ", result, "\n"))
   }
  }
library(plyr)
  library(tm)
  
source("wordBag.R")
locWords <- wordBag(locations, "locWords")
locationTable <- read.table("data/Location_Tree.csv", header=FALSE, sep="~", quote="", stringsAsFactors=FALSE,fill=TRUE)
locationTable$V1 <- gsub('"', "", locationTable$V1)
locationTable$V4 <- gsub('"', "", locationTable$V4)
  
} #end of lojack test function


#creates a wordBag of location name
createLocationWordBag <- function() {
  library(plyr)
  library(tm)
 
  source("wordBag.R")
  locationTable <- read.table("data/Location_Tree.csv", header=FALSE, sep="~", quote="", stringsAsFactors=FALSE,fill=TRUE)
  locationTable$V1 <- gsub('"', "", locationTable$V1)
  locationTable$V4 <- gsub('"', "", locationTable$V4)
  
 
  wordBags <- vector()
  i <- 1
  while(i < nrow(locationTable)){
    high <- i + 100;
    if(high > nrow(locationTable)){
      high <- nrow(locationTable)
    }
    wBag <- wordBag(locationTable[i:high,], "locWords",doStem=FALSE)
    wordBags <- union(wordBags, wBag)
    i <- i + 101
    cat(c("Finished with ", i/nrow(locationTable), "%\n"))
  }
  
  print(wordBags)
  
  locWords <- wordBags
  
#locWords <- wordBag(locationTable, "locWords")
save(locWords, file="data/locationWordBag.rda")
}#end of createLocationWordBag


#creates the place name to coordinate look up table
createCoordinateTable <- function() {
coordTable <- read.table("data/STRATEGI_2010_GAZETTEER.TXT", header=FALSE, sep="*", quote="", stringsAsFactors=FALSE)
cat("Saving coordinate table\n")
save(coordTable, file="data/coordinateTable.rda")
#table gives place name, county, and northing and easting
return(coordTable)
}#end of createCoordinateTable

#create lookup table for location texts
createLocationTable <- function() {
locationTable <- read.table("data/Location_Tree.csv", header=FALSE, sep="~", quote="", stringsAsFactors=FALSE,fill=TRUE)
locationTable$V1 <- gsub('"', "", locationTable$V1)
locationTable$V2 <- gsub('"', "", locationTable$V2)
locationTable$V3 <- gsub('"', "", locationTable$V3)
locationTable$V4 <- gsub('"', "", locationTable$V4)
locationTable$V5 <- gsub('"', "", locationTable$V5)

cat("Saving location table\n")
save(locationTable, file="data/locationTable.rda")
return(locationTable)
}#end of createLocationTable


coordinateCheck <- function(input) {

  load("data/coordinateTable.rda")
  name <- ""
  j <- ncol(input)
  name <- input[,j]
  while(name == ""){
    j <- j - 1
    name <- input[,j]
  }
 
results <- coordTable[coordTable$V1 == "zzzz",]
  while(nrow(results) == 0) {
    name <- input[,j]
cat(c(name, "\n"))
    results <- coordTable[coordTable$V1 == name,]
    print(results)
    j <- j - 1
  }

  if(nrow(results) == 1) {
    return("okay");
  }
  return(name)
} #end of coordinate check


wordBag.R
#################################
# wordBag.R
# written by Jet Goodson 
# started on 7 March 2013
# contact at jetgoodson@gmail.com
#
# makes a wordBag - intend mostly
# for analysis of common words
# not for vocab indexing
##################################


#makes a wordbag
wordBag <- function(input, bagName, doStem = TRUE){

  library("tm")

  bagOfHolding <- Corpus(DataframeSource(input))
  summary(bagOfHolding)
  bagOfHolding <- tm_map(bagOfHolding, removeNumbers)
  bagOfHolding <- tm_map(bagOfHolding, removePunctuation)
  bagOfHolding <- tm_map(bagOfHolding, stripWhitespace)
  bagOfHolding <- tm_map(bagOfHolding, tolower)
  bagOfHolding <- tm_map(bagOfHolding, removeWords, stopwords("english"))
  if(doStem==TRUE){
    bagOfHolding <- tm_map(bagOfHolding, stemDocument, language = "english")
  }
  terms <- TermDocumentMatrix(bagOfHolding)
  #matrixTerms <- as.matrix(terms)
  #matrixTerms <- rowSums(matrixTerms[,-1])
  return(findFreqTerms(terms, 1))
  
}#end of wordbag

#makes a text file for wordbagging
makeTextFile <- function(input, textFile){

  library("tm")
  write.table(input, file=textFile, row.names=FALSE, col.names=FALSE)

}#end of wordbag


#function to take a dataframe of key words and turn the key words into features.
featureBag <- function(frame, killStops = FALSE) {
  library('tm') 
  library('doMC')
  library('plyr')
  
  if ( require("multicore", quietly = FALSE, warn.conflicts = TRUE) ) {
    cat(c(multicore:::detectCores(), " available cores\n"))
    registerDoMC(cores = multicore:::detectCores())
  }
  
  freqMat <- data.frame()
  
  split <- 25000
  max <- nrow(frame)
  lows <- seq(1, max, by=split)
  highs <- vector()
  for(m in 1:length(lows)){
  highs <- c(highs, lows[m] + split - 1)  
  }
  if(highs[length(highs)] > max){
    highs[length(highs)] <- max
  }
    
  maxIt <- 1 + as.integer((nrow(frame) - 1)/split)
 
  freqMat <- foreach(v = 1:maxIt, .combine=rbind.fill) %dopar% {
    cat(c("Word bag", v, " starting on", lows[v], " to ", highs[v], "\n"))
    wordBag <- Corpus(DataframeSource(frame[lows[v]:highs[v],]))
    if(killStops == TRUE){
      wordBag <- tm_map(wordBag, removeNumbers)
      wordBag <- tm_map(wordBag, removePunctuation)
      wordBag <- tm_map(wordBag, stripWhitespace)
      wordBag <- tm_map(wordBag, tolower)
      wordBag <- tm_map(wordBag, removeWords, stopwords("english"))
    }
    wordBag <- TermDocumentMatrix(wordBag)
    wordBag <- as.matrix(wordBag)
    cat(c("Word bag", v, " finished\n"))
    wordBag <- t(wordBag)
    wordBag <- as.data.frame(wordBag)
  }#end of foreach loop

  return(freqMat)
}#end of feature bag



wagesOfSin.R
#################################
# wagesOfSin.R
# written by Jet Goodson 
# started on 7 March 2013
# contact at jetgoodson@gmail.com
#
# control file for wagesOfSin package
##################################

wagesOfSin <- function() {

  ###########################################################configure
  nEntries         <- 100000  #number of training entries to run over 
  proportion       <- 80/100   #split for train and validation
  samplingSize     <- 1000    #how many events is each tree made of; can save processing memory and time

  freeTextFeatures <- 400
  numFolds         <- 2
  numRepeats       <- 5
  tuneNum          <- 10
  numTrees         <- 500

  analyzeData      <- TRUE   #run analyzer on data; produces histograms of salary distribution vs word
  
  outputLog        <- "logFile.dat"
   
  predictionFileRF   <- "test_RF_salaryOutput.csv"
  predictionFileSVM   <- "test_SVM_salaryOutput.csv"
  
  loadSavedTrainData <- FALSE #load training data from saved instead of csv
  loadSavedTestData <- FALSE #load training data from saved instead of csv
  
  ################################################end of configuration
  sink(outputLog, append=TRUE, split=TRUE)
  
  
  library("tm")        #data mining library for nlp
  library("caret")
  library("SparseM")
  source("toolBox.R")
  source("wordBag.R")
  source("randomOrchard.R")
  source("cogInTheMachine.R")
   
 
  
  if(loadSavedTrainData == FALSE) {
    trainFrame <- read.csv("data/Train_rev1.csv", header=TRUE, skip=0, stringsAsFactors=FALSE, nrows=nEntries)
    cat(c(nrow(trainFrame), " rows in this sucker.\n"))
    
    
    trainFrame <- trainFrame[,!(colnames(trainFrame) %in% c("Id", "SalaryRaw", "LocationRaw"))] #not keen on these right now, maybe later, 
    
    trainFrame <- killDoppelgangers(trainFrame) #get rid of duplicate entries

    
    trainFrame <-cbind(trainFrame$SalaryNormalized, trainFrame[,!(colnames(trainFrame) %in% c("SalaryNormalized"))]) 

  
    goodWords <- vector()
    featureListing <- cullFreeText(trainFrame$Title, wordQuota = freeTextFeatures, goodList = goodWords)
    trainFrame$Title <- featureListing[[1]]
    goodWords <- featureListing[[2]]
     
 
    trainFrame <- transmogrifyFrame(trainFrame)
    
   # print(head(trainFrame))

cat(c("there are ", ncol(trainFrame), " columns now\n"))
    
    colnames(trainFrame)[1] <- "SalaryNormalized" #seems to get messed up

    trainFrame[is.na(trainFrame)] <- 0  #they seem to pop up - memory issue?
    
   # row.has.na <- apply(trainFrame, 1, function(x){any(is.na(x))})
   # row.has.na2 <- row.has.na[row.has.na == TRUE]
   # cat(c(length(row.has.na2), " number of NA rows\n"))
   # print(head(trainFrame[row.has.na, ]))

  
    
    save(trainFrame, file="data/table_train.rda")   #save so we can use the smaller version
    cat("Saved training data as R-object\n")    
  }
  if(loadSavedTrainData == TRUE) {   #load the saved bit
    cat("Loading:\n")
    print(load("data/table_train.rda"))
    cat("Loaded\n")
  }

  ####get and configure test data
  if(loadSavedTestData == FALSE){
  testFrame <- read.csv("data/Valid_rev1.csv", header=TRUE, skip=0, stringsAsFactors=FALSE)
  jobIdents <- testFrame[,1] #save job ids for later

  testFrame$Title <- cullFreeText(testFrame$Title, goodList = goodWords)[[1]]
  testFrame <- transmogrifyFrame(testFrame)
  testFrame[is.na(testFrame)] <- 0  #they seem to pop up - memory issue?
    
  save(testFrame, jobIdents, file="data/table_test.rda")
  }
  if(loadSavedTestData == TRUE) {   #load the saved bit
    cat("Loading:\n")
    print(load("data/table_test.rda"))
    cat("Loaded\n")
  }

   ###########################do some analysis
  
   if(analyzeData == TRUE){
     analyzer(trainFrame, "featurePlots")
   }

   ######################################split training data into a training set and a performance validation set
  set.seed(1)
  splitdex <- createDataPartition(trainFrame[,1], p=proportion, list=FALSE)
  validationFrame <- trainFrame[-splitdex,]
  trainFrame <- trainFrame[splitdex, ]
  cat(c("Training set partitioned into ", nrow(trainFrame), " for training and ", nrow(validationFrame), " for validation\n"))
  
   ################################################################## Random Forest 
 
  rfModel <- randomOrchard(trainFrame,
                           folds=numFolds,
                           repetitions=numRepeats,
                           tuneCount = tuneNum,
                           treeCount = numTrees)
  
  validateOrchard(rfModel, validationFrame)

  orchardOutput <- orchardPredict(rfModel, testFrame, jobIdents)
  write.csv(orchardOutput, file=predictionFileRF, row.names=FALSE,quote=FALSE)
  cat(c("Wrote predictions to file ", predictionFileRF, "\n"))


  ########################################################### SVM

#  svmModel <- cogInTheMachine(trainFrame)

 # validateSVM(rfModel, validationFrame)


  #svmOutput <- svmPredict(rfModel, testFrame, jobIdents)
  #write.csv(svmOutput, file=predictionFileSVM, row.names=FALSE,quote=FALSE)
  #cat(c("Wrote predictions to file ", predictionFileSVM, "\n"))


  
  return("Stipendium peccati mors est.")
} #end of wages of sin


randomOrchard.R
#################################
# randomOrchard.R
# written by Jet Goodson 
# started on 15 March 2013
# contact at jetgoodson@gmail.com
#
# handling of random forest alg on data
##################################

#set up and run random forest
randomOrchard <- function(trainFrame, samplingSize = 1000, folds = 2, repetitions = 5, tuneCount = 10, treeCount = 500) {

  library("randomForest")
  library("caret")
  library('doMC')
  
  #samplingSize     <- 1000    # these arguments are commented out because I moved them to function arguments

  rfMethod         <- "parRF"         #parRF == parallel random forest, in line is just rf
  validationMethod <- "repeatedcv"    #validation method for model
 # folds            <- 2              #number of folds in cv
 # repetitions      <- 1              #repetition for repeatedcv
 # tuneCount        <- 3              #number of tunings for mtry
 # treeCount        <- 500

  trainController <- trainControl(
    method = validationMethod,
    number=folds,
    repeats=repetitions,
    returnResamp = "final",
    classProbs = FALSE,
    returnData = FALSE
    )

  if ( require("multicore", quietly = FALSE, warn.conflicts = TRUE) ) {
    trainController$workers <- multicore:::detectCores()
    trainController$computeFunction <- mclapply
    trainController$computeArgs <- list(mc.preschedule = FALSE, mc.set.seed = FALSE)
    cat(c(multicore:::detectCores(), " available cores\n"))
    
    registerDoMC(cores = multicore:::detectCores())
  }
  
  cat("Training the orchard\n")
  
  theOrchard <- train(trainFrame[,-1], trainFrame[,1], method = "parRF", tuneLength = tuneCount, trControl = trainController, scale = FALSE, keep.forest=TRUE, sampsize=samplingSize, nTree = treeCount, na.action=na.omit)

  cat("Orchard trained\n")
  
  return(theOrchard)
}#end of random orchard

#run model on validationFrame and check accuracy
validateOrchard <- function(model, validationData) {
  library("randomForest")
  library("caret")

  cat("Validating RF:\n")
  
  validate <-predict(model, validationData[,-1])
  cat("RF Predicted head: \n")
  print(head(validate))

  cat("RF Actual head:\n")
  print(head(validationData[,1]))

  disparity <- validationData[,1] - validate

  cat(c("   RF Total Disparity = ", sum(abs(disparity)), "\n"))
  cat(c("   RF Average Disparity = ", sum(abs(disparity))/length(validate), "\n"))
  cat(c("   RF SD of Disparity = ", sd(abs(disparity)), "\n"))

  png(file="hist_salaryDisparity_RF.png")
  hist(disparity, main="RF Disparity, Actual - Prediction", xlab="Disparity", ylab="entries", col="darkorchid4")
  dev.off()
  
  png(file="hist_salaryRelDisparity_RF.png")
  hist(disparity/validationData[,1], main="RF Relative Disparity, (Actual - Prediction)/Actual", xlab="Relative Disparity", ylab="entries", col="darkorange1")
  dev.off()
  
  png(file="scatter_predictionVSactual_salary_RF.png")
  plot(validationData[,1], validate, main="RF Predicted Salary versus Actual Salary", xlab="Actual Salary", ylab="Predicted Salary",col="cornflowerblue")
  dev.off()

  cat("Finished RF validation.\n")

}#end of validateOrchard




#run the model over test data to produce predition
orchardPredict <- function(model, testFrame, jobIds){

  testPredict <- predict(model, newdata=testFrame, type="raw")
  cat("RF Prediction finished\n")
  results <- cbind(jobIds, testPredict)
  colnames(results) <- c("Id", "SalaryNormalized")
 
}#end of orchard predict


toolBox.R
#################################
# toolbox.R
# written by Jet Goodson 
# started on 7 March 2013
# contact at jetgoodson@gmail.com
#
# general data handling tools
##################################

#simplify and manipulate the fields
transmogrifyFrame <- function(dataFrame) {

  source("wordBag.R")
  library(prettyR)
  library(plyr)
  library(tm)
  library(Snowball)
  require(stringr)
  stopWords <- stopwords("en")
  class(stopWords)
  
  load(file="data/locationWordBag.rda")#gives locWords

  #commented out, lets let frequency count handle this
 # redundantWordBag <- getWordList(dataFrame$Category, c("&", "Jobs", "/", "  "), c("", "", " ", " ")) #this must be before next line to be useful
 # redundantWordBag <- c(redundantWordBag, "permanent", "full", "part", "time", "contract", "engineer", "financial", "recruiter", "teacher")
  redundantWordBag <- c("contract", "permanent", "part", "full", "time")
  dataFrame$Category <- transmogrifyCategory(dataFrame$Category)
  #want to remove accounting, nursing from list

  ###do location generalized
  dataFrame$LocationNormalized <- locateThis(dataFrame$LocationNormalized)

  #better column names
  dataFrame$ContractType <- gsub("part_time" , "partTimer", dataFrame$ContractType)   
  dataFrame$ContractType <- gsub("full_time" , "fullTimer", dataFrame$ContractType)
  #leave as words the ContractTime column

 
  dataFrame <- cbind(dataFrame[,!(colnames(dataFrame) %in% c("Title", "Category", "ContractTime", "ContractType", "LocationNormalized"))], featureBag(dataFrame[,(colnames(dataFrame) %in% c("Title", "Category", "ContractTime", "ContractType", "LocationNormalized"))]))

  #this is where the hard part starts
  
  #dataFrame$Title <- textCleanup(dataFrame$Title, c(stopWords, locWords, redundantWordBag))
  dataFrame$FullDescription <- textCleanup(dataFrame$FullDescription, c(stopWords, locWords, redundantWordBag))

  dataFrame$Company <- killPunkSpaces(dataFrame$Company)
  dataFrame$SourceName <- killPunkSpaces(dataFrame$SourceName)
  
#  cat(c("The company  column has ", length(freq(dataFrame$Company)[[1]]), " unique entries\n"))
#  cat(c("The source name  column has ", length(freq(dataFrame$SourceName)[[1]]), " unique entries\n"))

  dataFrame <- dataFrame[,!(colnames(dataFrame) %in% c("FullDescription", "SourceName", "Company"))]

  dataFrame <- dataFrame[,!(colnames(dataFrame) %in% c("nana", "full", "part", "time", "jobs"))] #not sure why these stick around
  
  return(dataFrame)
}#end of transmogrify

#getCategories - word bag the categories
getWordList <- function(column, find = c(""), replace = c("")){
  if(length(find) > 0 && length(find) == length(replace)){
    for(i in 1:length(find)) {
      column <- gsub(find[i], replace[i], column)
    }}
  if(length(find) != length(replace)){
    cat("Find and replace lengths do not match\n")
  }
  column <- gsub("Nursing", "other", column) #for later redundancy, want to distinguish accountants from financiers, nurses from doctors
  column <- gsub("Accounting", "other", column)
  wordList <- tolower(unique(unlist(strsplit(column, " "))))
  return(wordList)
} #end of getCategories



#specifically alters 
transmogrifyCategory <- function(column) {
  #I'm turning this into [hopefully] unique identifiers that will make good feature names in columns. And being smartassed.
  column <- gsub("Accounting & Finance Jobs", "fatCats", column)  
  column <- gsub("Admin Jobs", "theMan", column)
  column <- gsub("Charity & Voluntary Jobs", "bleedingHearts", column)
  column <- gsub("Consultancy Jobs", "hiredGuns", column)
  column <- gsub("Creative & Design Jobs", "starvingArtists", column)
  column <- gsub("Customer Services Jobs", "helpDesk", column)
 #column <- gsub("Domestic Help & Cleaning Jobs", "custodialEngineering", column)
  column <- gsub("Domestic help & Cleaning Jobs", "custodialEngineering", column)
  column <- gsub("Energy, Oil & Gas Jobs", "goJuice", column)
  column <- gsub("Engineering Jobs", "beamMeUpScotty", column)
  column <- gsub("Graduate Jobs", "poorBastards", column)
  column <- gsub("Healthcare & Nursing Jobs", "bonesMcCoy", column)
  column <- gsub("Hospitality & Catering Jobs", "normanBates", column)
  column <- gsub("HR & Recruitment Jobs", "headHunters", column)
  column <- gsub("IT Jobs", "nerdHerd", column)
  column <- gsub("Legal Jobs", "legalEagles", column)
  column <- gsub("Logistics & Warehouse Jobs", "inTheRearWithTheGear", column)
  column <- gsub("Maintenance Jobs", "mrFixit", column)
  column <- gsub("Manufacturing Jobs", "assemblyLine", column)
  column <- gsub("Other/General Jobs", "oddJob", column)
  column <- gsub("PR, Advertising & Marketing Jobs", "prFlacks", column)
  column <- gsub("Property Jobs", "groundkeeperWillie", column)
  column <- gsub("Retail Jobs", "poshShopGirl", column)
  column <- gsub("Sales Jobs", "usedCarSalesmen", column)
  column <- gsub("Scientific & QA Jobs", "madScientists", column)
 #column <- gsub("Social Work Jobs", "busyBodies", column)
  column <- gsub("Social work Jobs", "busyBodies", column)
  column <- gsub("Teaching Jobs", "glorifiedBabysitters", column)
  column <- gsub("Trade & Construction Jobs", "doozers", column)
  column <- gsub("Travel Jobs" , "leavinOnAJetPlane", column)
  return(column)
}#end of transmogrifyCategory



#takes location to the second column to minimize features, should be about 30
locateThis <- function(column) {
  load(file="data/locationTable.rda")#gives locWords
  column <- sapply(column, function(x){
    newLoc <- which(locationTable$V4 == x)
    if(length(newLoc) == 0){
      newLoc <- which(locationTable$V3 == x)
      if(length(newLoc) == 0){
        newLoc <- x 
      }}
    if(length(newLoc) > 0){
     newLoc <- locationTable[newLoc[1], 2] 
    }
    #newLoc <- tolower(newLoc)
    newLoc <- strsplit(newLoc, " ")[[1]]
    newLoc <- paste(toupper(substring(newLoc, 1, 1)), substring(newLoc, 2), collapse = "")
  })
  column <- gsub("/" , "", column)
  column <- gsub(" " , "", column)
  column <- gsub("&" , "", column)
  column <- gsub("uk" , "", column) #duh. If there are any ads outside UK, this field won't tell us
  return(column) 
} #end of locateThis


#kill punctuation and spaces
killPunkSpaces <- function(column) {
  column <- gsub("[[:punct:]]", " ", column)  #kill punctuation, do before following to preserve urls in sourcename
  column <- sapply(column, function(x){
    t <- strsplit(x, " ")[[1]]
    t <- paste(toupper(substring(t, 1, 1)), substring(t, 2), sep = "", collapse = "")
  })
  column <- gsub("/\\s/", "", column) #strip excess white space
}#end of killPunkSpaces


#removes words from column that appear in other category, i.e. locations and job categories
wordRemover <- function(column, wordList) {
  '%nin%' <- Negate('%in%')
  column <- lapply(column, function(x) {   
    t <- unlist(strsplit(x, " "))
    t <- t[t %nin% wordList] ##this gets rid of stopwords i.e., common words like article, simple verbs
    t <- paste(t, sep=" ", collapse= " ") 
  })
  return(column)
}#end of word remover


#do all the things to the free text in title and fulldescription, saved from how it was around March 10th, 2013 for reference
textCleanup <- function(column, killWords = c("")) {
  source("wordBag.R")
  
  library(plyr)
  library(tm)
  library(Snowball)
  require(stringr)
  
  column <- tolower(column)   #make it lower case
  
  column <- gsub("[_,\\*]", "", column)  ##get rid of where they replaced salaries with asterisks.
                                        #below gets rid of websites, which I don't think we need to keep, they mess things up
  column <- gsub("(((file|gopher|news|nntp|telnet|http|ftp|https|ftps|sftp)://)|(www\\.))+(([a-zA-Z0-9\\._-]+\\.[a-zA-Z]{2,6})|([0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}))(/[a-zA-Z0-9\\&amp;%_\\./-~-]*)?", "", column)
  column <- gsub("^.*(.com|.co.uk).*$", "", column)   #destroy british websites
  column <- gsub("/", " ", column)    #get rid of slashes, confuses things
  column <- gsub("\\.", " ", column)  #get rid of periods
  
  
  '%nin%' <- Negate('%in%')
  column <- lapply(column, function(x) {   
    t <- unlist(strsplit(x, " "))
    #  t <- t[t %nin% stopWords] ##this gets rid of stopwords i.e., common words like article, simple verbs
    t <- t[t %nin% killWords]#this gets rid of locations
    t <- SnowballStemmer(t)
    t <- paste(t, sep=" ", collapse= " ") 
  })

  column <- gsub("[[:punct:]]", " ", column)  #kill punctuation, do before following to preserve urls in sourcename
  column <- gsub("\\d", " ", column)  #get rid of digits
  column <- gsub("/\\s\\s+/", " ", column) #strip excess white space
  column <- str_trim(column)
  
  return(column)
}#end of text cleanup

#get rid of repeated entries in dataset
killDoppelgangers <- function(frame){
  clones <- duplicated(frame)
  clones <- which(clones == TRUE)
  if(length(clones) < 1){
    return(frame)
  }
  cat(c("There are ", nrow(frame[clones,]), " duplicate entries\n"))
  return(frame[-clones, ])
}#end of doppelganger function


#produce some plots
analyzer <- function(frame, folder){
  system(paste(c("mkdir -p ", folder), collapse=""))
  par(bg = "white")
  colNames <- colnames(frame)

  for(i in 2:length(colNames)){
    png(file=paste(c(folder, "/histogram_", colNames[1], "_for_", colNames[i], ".png"), collapse=""), width=800,height=400)
    hist(frame[(frame[,i]==1),1], main=paste(c(colNames[1], " for ", colNames[i]),collapse=""), xlab=paste(c(colNames[1], " [Pounds] "), collapse=""), ylab="entries",col="red3") 
  dev.off()
  }#end of histogram loop

  png(file="hist_salaryDist.png")
  hist(frame[,1], main="Salary Distribution", xlab=paste(c(colNames[1], " [Pounds] "), collapse=""), ylab="entries",col="cornflowerblue") 
  dev.off()
  
  graphics.off()
  
}#end of analyzer


#function to pre-process free text (title and fulldescription) for feature creation by eliminating least common words
cullFreeText <- function(column, wordQuota = 400, goodList = vector(), folder="wordPlots", doPlots=FALSE){

  system(paste(c("mkdir -p ", folder), collapse=""))
  
  library('plyr')
  library(Snowball)
  require(stringr)
  load(file="data/locationWordBag.rda")#gives locWords
  stopWords <- stopwords("en")
  class(stopWords)
  
  column <- tolower(column)   #make it lower case
  
  '%nin%' <- Negate('%in%')
  column <- sapply(column, function(x) {   
    t <- unlist(strsplit(x, " "))
    t <- t[t %nin% locWords]#this gets rid of locations
    t <- paste(t, sep=" ", collapse= " ") 
  })
  
  column <- gsub("\\+\\+", "plusplus", column)    #want to preserve C skills when 
  column <- gsub("\\#", "sharp", column)          #we kill special characters
  column <- gsub("c/c", "ceecee", column)    
  
  column <- gsub("/", " ", column)    #get rid of slashes, confuses things
  column <- gsub("\\.", " ", column)  #get rid of periods
  column <- gsub("\\d", "", column)
  column <- gsub("/\\s\\s+/", " ", column) #strip excess white space
  
  '%nin%' <- Negate('%in%')
  column <- sapply(column, function(x) {   
    t <- unlist(strsplit(x, " "))
    t <- t[t %nin% stopWords] ##this gets rid of stopwords i.e., common words like article, simple verbs
    t <- SnowballStemmer(t)
    t <- unique(t)
    t <- paste(t, sep=" ", collapse= " ") 
  })
  
  column <- gsub("[[:punct:]]", " ", column)  #kill punctuation, do before following to preserve urls in sourcename
  column <- gsub("\\b[a-zA-Z0-9]{1,2}\\b", "", column) #remove single characters
  column <- gsub("/\\s\\s+/", " ", column) #strip excess white space
  
  
  wordList <- sapply(column, function(x) {
    t <- unlist(strsplit(x, " ")) 
  })
  
  wordList <- unlist(wordList)
  freqWords <- sort(table(wordList), decreasing=TRUE)
  
  if(doPlots == TRUE){
  png(file=paste(c(folder, "/", "hist_unstructTextWordFrequencies_High.png"), collapse=""))
  hist(freqWords[freqWords > 5000], main = "freq > 5000 Occurances", col="cornflowerblue")
  dev.off()
  
  
  png(file=paste(c(folder, "/", "hist_unstructTextWordFrequencies_Low.png"), collapse=""))
  hist(freqWords[freqWords < 5000], main = "Freq < 5000 Occurances", col="cornflowerblue")
  dev.off()
  
  png(file=paste(c(folder, "/", "hist_unstructTextWordFrequencies_ReallyLow.png"), collapse=""))
  hist(freqWords[freqWords < 1000], main = "Freq < 1000 Occurances", col="cornflowerblue")
  dev.off()
  
  png(file=paste(c(folder, "/", "hist_unstructTextWordFrequencies_ReallyVeryLow.png"), collapse=""))
  hist(freqWords[freqWords < 100], main = "Freq < 100 Occurances", col="cornflowerblue")
  dev.off()
  
  png(file=paste(c(folder, "/", "hist_unstructTextWordFrequencies.png"), collapse=""))
  hist(freqWords[-1], main = "Word count, All Freq", col="cornflowerblue", breaks=c(0,1,10,100,1000,5000,10000,15000,20000,25000,50000,100000))
  dev.off()
  
  png(file=paste(c(folder, "/", "hist_unstructTextWordFrequencies_firstHundred.png"), collapse=""))
  hist(freqWords[2:(wordQuota+1)], main = "Freq of Quota Most Common Words", col="cornflowerblue")
  dev.off()
  
  png(file=paste(c(folder, "/", "hist_unstructTextWordFrequencies_ThousandBreaks.png"), collapse=""))
  hist(freqWords[-1], main = "Word Count, All Freq (1000 bins)", col="cornflowerblue", breaks=1000)
  dev.off()
  

  png(file=paste(c(folder, "/", "hist_unstructTextWordFrequencies_handfuls.png"), collapse=""))
  hist(freqWords[freqWords < 10], main = "Freq < 10 Occurances", col="cornflowerblue")
  dev.off()
}

 
  if(length(goodList) > 0){
    killList <- rownames(freqWords)
    killList <- setdiff(killList, goodList)
  }

  if(length(goodList) == 0){
    killList <- rownames(freqWords[(wordQuota+2):length(freqWords)])
    killList <- sort(unique(killList))
    goodList <- rownames(freqWords[2:(wordQuota+1)])
  }
  
  '%nin%' <- Negate('%in%')
  column <- sapply(column, function(x) {   
    t <- unlist(strsplit(x, " "))
    t <- t[t %nin% killList]#this gets rid of locations
    t <- paste(t, sep=" ", collapse= " ") 
  })

  column <- gsub("/\\s\\s+/", " ", column) #strip excess white space -- it's persistent

  return(list(column, goodList))
  
}#end of pretreat text


#gets a SparseM sparse matrix from a data frame
getSparse <- function(dataFrame) {
    tmpMat <- as.matrix(dataFrame)
    sparseMat <- as.matrix.csr(tmpMat)
    rm(tmpMat)
    return(sparseMat)
} #end of getSparse



